File: 189-RotateArray/189-RotateArray.py
# Last updated: 9/14/2025, 9:16:08 AM
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        k=k%len(nums)
        nums.reverse()
        start,end=0,k-1
        while(start<end):
            nums[start],nums[end]= nums[end],nums[start]
            start+=1
            end-=1
        start,end=k,len(nums)-1
        while(start<end):
            nums[start],nums[end]= nums[end],nums[start]
            start+=1
            end-=1

File: 189-RotateArray/189-RotateArray.java
// Last updated: 9/14/2025, 9:16:00 AM
class Solution {
    public void rotate(int[] nums, int k) {
        int  len = nums.length;
        k = k % len;
        if(k == 0) return;
        k--;
        reverse(nums, 0, nums.length-1);
        reverse(nums, 0, k);
        reverse(nums, k+1, nums.length-1);
    }
    public static void reverse(int []arr, int start, int end){
        while(start < end){
            swap(arr, start, end);
            start++;
            end--;
        }
    }
    public static void swap(int []arr, int x, int y){
        int temp = arr[x];
        arr[x] = arr[y];
        arr[y] = temp;
    }
}